### PR TITLE
#163519098 Implement forgot password page using React

### PR DESCRIPTION
**Pull request by oesukam**
#### What does this PR do?
Implement forgot password page using React

#### Description of Task to be completed?
- Write tests for redux
- Write tests for App component
- Add Button as a common component
- Add Input as a common component
- Add ForgotPasswowd component
- Add ResetPassword component
- Add project's information in `README.md`

#### How should this be manually tested?
- Clone the repository
- Run `npm run start`
- Navitage to `/forgot-password` url using any browser
- Fill and submit the form
- Check your inbox for the reset password link
- Click on the reset password link
- Fill and submit your new password

#### What are the relevant pivotal tracker stories?
`#163519098`

#### Screenshots (if appropriate)
image

image

image

[Test coverages](https://codeclimate.com/repos/5ca1a7a152587c02850004cf/settings/test_reports/5caa0740335e6d0286002ee1)

image

#### Questions:
N/A

### CODE REVIEW

File: src/components/ForgotPassword/ForgotPassword.jsx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+export class ForgotPassword extends Component {
+  constructor() {
+    super();
+    this.state = {
+      validEmail: true,
+      emailError: 'Email is not valid',
+    };
+    this.handleInputChange = this.handleInputChange.bind(this);
+    this.onSubmitButton = this.onSubmitButton.bind(this);
+    this.renderErrors = this.renderErrors.bind(this);
```
Comment (kimotho-njoki) [inline src/components/ForgotPassword/ForgotPassword.jsx]: I'm quite skeptical about this bind statements. Why do you need to add them? If it's an eslint error please send me the error.

File: src/components/ForgotPassword/ForgotPassword.jsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    e.preventDefault();
+  }
+
+  handleInputChange(e) {
+    this.setState({ validEmail: true });
+    const { onInputChange } = this.props;
+    onInputChange({ field: e.target.name, value: e.target.value });
+  }
+
+  renderErrors() {
```
Comment (kimotho-njoki) [inline src/components/ForgotPassword/ForgotPassword.jsx]: Why are we not making use of ES6 ?

File: src/components/ForgotPassword/ResetPassword.jsx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    } = this.props;
+    this.setState({ validPassword: true });
+    if (newPassword.length < 5 || confirmNewpassword.length < 5) {
+      this.setState({
+        validPassword: false,
+        passwordError: 'Password must have at least 6 chacters',
+      });
+      return;
+    }
+    if (newPassword !== confirmNewpassword) {
```
Comment (kimotho-njoki) [inline src/components/ForgotPassword/ResetPassword.jsx:34]: I thought these checks were all handled in backend?

File: src/components/__tests__/components/common/Input.test.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import { shallow } from 'enzyme';
+import Input from '../../../common/Input/Input';
+
+describe('<Input />', () => {
+  test('Should render the <Input />', () => {
+    const handleInput = () => true;
+    const wrapper = shallow(<Input onChange={handleInput} />);
+    expect(wrapper).toMatchSnapshot();
```
Comment (kimotho-njoki) [inline src/components/__tests__/components/common/Input.test.js:9]: Are you all using snapshots to test?

File: src/components/ForgotPassword/ForgotPassword.jsx (hunk 1 above)
Comment (oesukam) [reply src/components/ForgotPassword/ForgotPassword.jsx]: @kimotho-njoki  The Eslint rules from airbnb that we are using don't allow arrow functions. We are binding `this` in order to have access to props and state within those functions

image

File: src/components/ForgotPassword/ForgotPassword.jsx (hunk 2 above)
Comment (oesukam) [reply src/components/ForgotPassword/ForgotPassword.jsx]: @kimotho-njoki The Esling configuration that we are using does not allow arrow functions within a react component.
image

File: src/components/ForgotPassword/ResetPassword.jsx (hunk 3 above)
Comment (oesukam) [reply src/components/ForgotPassword/ResetPassword.jsx:34]: @kimotho-njoki We are handling in the both backend and frontend

Comment (oesukam) [reply src/components/ForgotPassword/ResetPassword.jsx:34]: @kimotho-njoki We are handling in both backend and frontend

File: src/components/__tests__/components/common/Input.test.js (hunk 4 above)
Comment (oesukam) [reply src/components/__tests__/components/common/Input.test.js:9]: @kimotho-njoki Yes, we are all using snapshots to test component rendering

File: src/components/ForgotPassword/ForgotPassword.jsx (hunk 1 above)
Comment (kimotho-njoki) [reply src/components/ForgotPassword/ForgotPassword.jsx]: Well yes but it still doesn't answer my question. When you fail to bind your statements do you get an error message. Have you tried not binding your statement??

Comment (kimotho-njoki) [reply src/components/ForgotPassword/ForgotPassword.jsx]: Have you tried only having state and eradicating the constructor?

Comment (kimotho-njoki) [reply src/components/ForgotPassword/ForgotPassword.jsx]: I thought we were using Es6... So if our eslint doesn't allow for that why not change the rules??? When using ES5 you will have to keep binding your statements.. Most projects don't do that anymore.. Es6 doesn't need binding.. Research more on that...

Comment (oesukam) [reply src/components/ForgotPassword/ForgotPassword.jsx]: @kimotho-njoki It is only an Eslint problem. I have changed Eslint rules to allow arrow functions and I also refactored all components accordingly.

**Review by mcaleb808 — APPROVED** [review]

**Review by filschristian — APPROVED** [review]

**Review by NkFab — APPROVED** [review]

File: src/components/__tests__/actions/forgotPasswordActions.test.js [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+const mockStore = configureStore([thunk]);
+let store;
+jest.setTimeout(30000);
+
+describe('forgotPassword', () => {
+  describe('synchronous actions', () => {
+    beforeEach(() => {
+      store = mockStore({});
+    });
+    test('should dispatch FORGOT_PASSWORD_SUCCESS action', () => {
```
Comment (kimotho-njoki) [inline src/components/__tests__/actions/forgotPasswordActions.test.js:31]: space this test

File: src/components/__tests__/actions/forgotPasswordActions.test.js [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        type: FORGOT_PASSWORD_FORM,
+        payload,
+      };
+
+      store.dispatch(handleForgotPasswordForm(payload));
+
+      const actions = store.getActions();
+      expect(actions).toEqual([expectedPayload]);
+    });
+    test('should dispatch FORGOT_PASSWORD_SUCCESS action', () => {
```
Comment (kimotho-njoki) [inline src/components/__tests__/actions/forgotPasswordActions.test.js:44]: space this test

File: src/components/__tests__/components/ResetPassword.test.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+      wrapper.find('Input[name="newPassword"]').simulate('change', { target: { value: '12345' } });
+    });
+    test('should update the newPassword prop', () => {
+      expect(wrapper.state()).toEqual(defaultState);
+    });
+    test('should update the newPassword prop', () => {
+      expect(wrapper.state()).toEqual(defaultState);
+    });
+  });
+  describe('when typing into the confirmNewpassword input', () => {
```
Comment (kimotho-njoki) [inline src/components/__tests__/components/ResetPassword.test.js:70]: space this test

File: src/redux/reducers/forgotPasswordReducer.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      };
+    case FORGOT_PASSWORD_FORM:
+      return {
+        ...state,
+        [payload.field]: payload.value,
+        errors: [],
+        message: '',
+      };
+    case FORGOT_PASSWORD_SUCCESS:
+      return { ...state, successMessage: payload.message, submitting: false };
```
Comment (kimotho-njoki) [inline src/redux/reducers/forgotPasswordReducer.js]: why is this return in a single line?

Comment (oesukam) [reply src/redux/reducers/forgotPasswordReducer.js]: @kimotho-njoki it was my auto format from VSCode putting it back to one line. I have disabled it and change to multiline.

File: src/components/__tests__/components/ResetPassword.test.js (hunk 7 above)
Comment (oesukam) [reply src/components/__tests__/components/ResetPassword.test.js:70]: @kimotho-njoki I have added an empty line before describe (line 65).

File: src/components/__tests__/actions/forgotPasswordActions.test.js (hunk 6 above)
Comment (oesukam) [reply src/components/__tests__/actions/forgotPasswordActions.test.js:44]: @kimotho-njoki I have added an empty line before test (line 42).

File: src/components/__tests__/actions/forgotPasswordActions.test.js (hunk 5 above)
Comment (oesukam) [reply src/components/__tests__/actions/forgotPasswordActions.test.js:31]: @kimotho-njoki I have added an empty line before test (line 30).

**Pull request merged**